File: src/demos/ps/oo/Job.java
package demos.ps.oo;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import demos.ps.oo.Item.ItemType;
import demos.ps.oo.PSLibSpell.PS1Spell;
import demos.ps.oo.PSLibSpell.Spell;

import static java.util.Arrays.asList;

// TODO Add each Job -> Initial Equipment and refactor defaultEquipment from Partymember

public enum Job {
	ADVENTURER 	(16, 4, 2.2, 2.7), // Alis, Rolf, Chaz 		36	Weap (9) + ATKTech(10) + ORITech (5) = 60
	NATURER 	(18, 4, 2.4, 2.4), // Nei, Ryka, Myau		39	Weap (6) + CURTech (10) + AGI (5) = 60
	FIGHTER		(24, 1, 3.1, 2.9), // Odin, Rudo			48	Weap(12)	=	60
	ESPER		(11, 11, 1.4, 2.2), // Noah, Rune			33	Weap (3) + ATK2Tech (18) + MiscTech (6) = 60
	ROBOT		(12, 1, 2.0, 2.6), // Hapsby
	// TODO Below this, review later
	HUNTER		(18, 3, 2.0, 2.5), // Anna, Alys, Kyra      39	Weap (6) + CBTTech(10) + AGI (5)
	MECHANIC	(21, 3, 2.0, 2.5), // Kain, Gryz			45	Weap (9) + AntiTech (6) = 60
	PRIEST 		(12, 10, 1.0, 2.5), // Raja					34	Weap (3) + CUR2Tech (18) + DEFTech (5) = 60 
	SCHOLAR 	(13, 8, 1.0, 2.5), // Hahn, Hugh			32	Weap (3) + CURTech (10) + CBTTech(10) + DEFTech (5) = 60
	HEALER 		(12, 9, 1.0, 2.5), // Amy					33	Weap (3) + CUR2Tech (18) + MiscTech (6) = 60
	THIEF		(15, 4, 1.0, 2.5), // Shir					34	Weap (3) + MiscTech (6) + AGI (5) + THI (12) = 60
	GUARDIAN	(20, 2, 2.0, 2.5), // Wren      			42	Weap (6) + MiscTech (6) + AutoCure (6) = 60
	WATCHER		(18, 2, 2.0, 2.5), // Demi					38	Weap (6) + CURTech (10) + AutoCure (6) = 60
	;

	private final int baseHp, baseMp;
	private final double aMod, dMod;

	private Map<Integer, Spell[]> map_level_spell;
	
	Job(int hp, int mp, double aMod, double dMod) {
		this.baseHp = hp;
		this.baseMp = mp;
		this.aMod = aMod;
		this.dMod = dMod;
	}

	public Map<Integer, Spell[]> getMapLevelSpell() {
		if(this.map_level_spell == null) {
			this.initLevelSpellMapping();			 			
		}
		return this.map_level_spell;
	}
	
	public int getHp(Specie spe, int level) {
		float value = baseHp + spe.modHp;
		for (int i = 2; i <= level; i++)
			value = value + ((float) (baseHp + spe.modHp) / 4)
					* (float) (i + 6) / 9;
		return Math.round(value);
	}

	public int getMp(Specie spe, int level) {
		float value = baseMp + spe.modMp;
		for (int i = 2; i <= level; i++) {
			value = value + ((float) (baseMp + spe.modMp) / 4)
					* (float) (i + 6) / 7;
		}
		value = Math.round(value);

		if (value < 20 && value % 2 == 1) // Make it even for small values
			return (int) (value + 1);
		else
			return (int) value;
	}

	public int listXp(int level) {
		int pertubation = (int) (2.3 * this.baseHp + 1.7 * this.baseMp);
		float value = 0;
		for (int i = 2; i <= level; i++)
			value = value + ((13) * (i * i) / 7) + 10;
		value = Math.round(value
				+ (int) ((pertubation * value / (190 - level))));
		return (int) value;
		//return (int) Math.min(value, 999999);
	}
	
	public int getAtk(int level) {
		double atk = (int) (Math.floor(aMod*4));
		for(int i=1;i<level;i++) {
			atk+= aMod;
		}
		return (int) Math.round(atk);
	}
			
	public int getDef(int level) {
		double def = (int) (Math.round(dMod*3));
		for(int i=1;i<level;i++) {
			def+= dMod;
		}
		return (int) Math.round(def);
	}
	
	public void initLevelSpellMapping() {

		this.map_level_spell = new HashMap<Integer, Spell[]>();
		switch(this) {
			case ADVENTURER:
				map_level_spell.put(new Integer(4), new Spell[]{PS1Spell.REST});
				map_level_spell.put(new Integer(5), new Spell[]{PS1Spell.ESCAPE});
				map_level_spell.put(new Integer(6), new Spell[]{PS1Spell.CHAT});
				map_level_spell.put(new Integer(8), new Spell[]{PS1Spell.FIRE}); // Changed, was 12
				map_level_spell.put(new Integer(14), new Spell[]{PS1Spell.ROPE});
				map_level_spell.put(new Integer(16), new Spell[]{PS1Spell.FLY});
				map_level_spell.put(new Integer(20), new Spell[]{PS1Spell.CURE}); // New
				map_level_spell.put(new Integer(26), new Spell[]{PS1Spell.GI_FIRE}); // New
				map_level_spell.put(new Integer(30), new Spell[]{PS1Spell.ROPE_ALL}); // New
				break;
			
			case NATURER:
				map_level_spell.put(new Integer(6), new Spell[]{PS1Spell.CURE});
				map_level_spell.put(new Integer(9), new Spell[]{PS1Spell.FEAR});
				map_level_spell.put(new Integer(12), new Spell[]{PS1Spell.WALL});
				map_level_spell.put(new Integer(15), new Spell[]{PS1Spell.TRAP});
				map_level_spell.put(new Integer(17), new Spell[]{PS1Spell.EXIT});
				map_level_spell.put(new Integer(20), new Spell[]{PS1Spell.FORCE});
				map_level_spell.put(new Integer(22), new Spell[]{PS1Spell.FEAR_ALL}); // New
				map_level_spell.put(new Integer(25), new Spell[]{PS1Spell.POWER_CURE}); // New
				break;

			case ESPER: // shifted by 2 some spells
				map_level_spell.put(new Integer(1), new Spell[]{PS1Spell.LIGHT, PS1Spell.FIRE});
				map_level_spell.put(new Integer(4), new Spell[]{PS1Spell.W_REST}); // New
				map_level_spell.put(new Integer(7), new Spell[]{PS1Spell.W_CURE});
				map_level_spell.put(new Integer(6+2), new Spell[]{PS1Spell.EXIT});
				map_level_spell.put(new Integer(8+2), new Spell[]{PS1Spell.TELE});
				map_level_spell.put(new Integer(12), new Spell[]{PS1Spell.WIND});
				map_level_spell.put(new Integer(14), new Spell[]{PS1Spell.PROT});
				map_level_spell.put(new Integer(16), new Spell[]{PS1Spell.GI_FIRE}); // New
				map_level_spell.put(new Integer(18), new Spell[]{PS1Spell.THUNDER});
				map_level_spell.put(new Integer(20), new Spell[]{PS1Spell.REVIVE});
				map_level_spell.put(new Integer(22), new Spell[]{PS1Spell.OPEN}); // Was 17
				map_level_spell.put(new Integer(28), new Spell[]{PS1Spell.F_REVIVE}); // New
				break;
		default:
			break;
		}
	}
	
	@Override
	public String toString() {
		String s = super.toString();
		return PSGame.getString("Job_" + s.substring(0, 1) + s.substring(1).toLowerCase());
	}
	
};	

File: src/demos/ps/game/Job.ts
/**
 * Job - Character Job System
 * Direct port of Job.java - Defines character jobs/classes with stats and spell progression
 */

import { PSGame } from '../PSGame';
import { Spell, PS1Spell, SpellFactory } from './PSLibSpell';
import { Specie, SpecieHelper } from './Specie';

// Job enum with stats - direct port from Job.java
export enum Job {
  ADVENTURER = 'ADVENTURER', // Alis, Rolf, Chaz - 36 Weap (9) + ATKTech(10) + ORITech (5) = 60
  NATURER = 'NATURER',       // Nei, Ryka, Myau - 39 Weap (6) + CURTech (10) + AGI (5) = 60
  FIGHTER = 'FIGHTER',       // Odin, Rudo - 48 Weap(12) = 60
  ESPER = 'ESPER',           // Noah, Rune - 33 Weap (3) + ATK2Tech (18) + MiscTech (6) = 60
  ROBOT = 'ROBOT',           // Hapsby
  HUNTER = 'HUNTER',         // Anna, Alys, Kyra - 39 Weap (6) + CBTTech(10) + AGI (5)
  MECHANIC = 'MECHANIC',     // Kain, Gryz - 45 Weap (9) + AntiTech (6) = 60
  PRIEST = 'PRIEST',         // Raja - 34 Weap (3) + CUR2Tech (18) + DEFTech (5) = 60
  SCHOLAR = 'SCHOLAR',       // Hahn, Hugh - 32 Weap (3) + CURTech (10) + CBTTech(10) + DEFTech (5) = 60
  HEALER = 'HEALER',         // Amy - 33 Weap (3) + CUR2Tech (18) + MiscTech (6) = 60
  THIEF = 'THIEF',           // Shir - 34 Weap (3) + MiscTech (6) + AGI (5) + THI (12) = 60
  GUARDIAN = 'GUARDIAN',     // Wren - 42 Weap (6) + MiscTech (6) + AutoCure (6) = 60
  WATCHER = 'WATCHER'        // Demi - 38 Weap (6) + CURTech (10) + AutoCure (6) = 60
}

export class JobHelper {
  private static readonly jobConfigs = new Map<Job, {
    baseHp: number, baseMp: number, aMod: number, dMod: number
  }>([
    [Job.ADVENTURER, { baseHp: 16, baseMp: 4, aMod: 2.2, dMod: 2.7 }],
    [Job.NATURER, { baseHp: 18, baseMp: 4, aMod: 2.4, dMod: 2.4 }],
    [Job.FIGHTER, { baseHp: 24, baseMp: 1, aMod: 3.1, dMod: 2.9 }],
    [Job.ESPER, { baseHp: 11, baseMp: 11, aMod: 1.4, dMod: 2.2 }],
    [Job.ROBOT, { baseHp: 12, baseMp: 1, aMod: 2.0, dMod: 2.6 }],
    [Job.HUNTER, { baseHp: 18, baseMp: 3, aMod: 2.0, dMod: 2.5 }],
    [Job.MECHANIC, { baseHp: 21, baseMp: 3, aMod: 2.0, dMod: 2.5 }],
    [Job.PRIEST, { baseHp: 12, baseMp: 10, aMod: 1.0, dMod: 2.5 }],
    [Job.SCHOLAR, { baseHp: 13, baseMp: 8, aMod: 1.0, dMod: 2.5 }],
    [Job.HEALER, { baseHp: 12, baseMp: 9, aMod: 1.0, dMod: 2.5 }],
    [Job.THIEF, { baseHp: 15, baseMp: 4, aMod: 1.0, dMod: 2.5 }],
    [Job.GUARDIAN, { baseHp: 20, baseMp: 2, aMod: 2.0, dMod: 2.5 }],
    [Job.WATCHER, { baseHp: 18, baseMp: 2, aMod: 2.0, dMod: 2.5 }]
  ]);

  // Level-to-spell mapping cache
  private static readonly spellMappings = new Map<Job, Map<number, Spell[]>>();

  public static getBaseHp(job: Job): number {
    return this.jobConfigs.get(job)?.baseHp || 0;
  }

  public static getBaseMp(job: Job): number {
    return this.jobConfigs.get(job)?.baseMp || 0;
  }

  public static getAMod(job: Job): number {
    return this.jobConfigs.get(job)?.aMod || 0;
  }

  public static getDMod(job: Job): number {
    return this.jobConfigs.get(job)?.dMod || 0;
  }

  // HP calculation - direct port from Java
  public static getHp(job: Job, specie: Specie, level: number): number {
    const baseHp = this.getBaseHp(job);
    const modHp = SpecieHelper.getModHp(specie);
    let value = baseHp + modHp;

    for (let i = 2; i <= level; i++) {
      value = value + ((baseHp + modHp) / 4) * (i + 6) / 9;
    }

    return Math.round(value);
  }

  // MP calculation - direct port from Java
  public static getMp(job: Job, specie: Specie, level: number): number {
    const baseMp = this.getBaseMp(job);
    const modMp = SpecieHelper.getModMp(specie);
    let value = baseMp + modMp;

    for (let i = 2; i <= level; i++) {
      value = value + ((baseMp + modMp) / 4) * (i + 6) / 7;
    }

    value = Math.round(value);

    // Make it even for small values
    if (value < 20 && value % 2 === 1) {
      return Math.floor(value + 1);
    } else {
      return Math.floor(value);
    }
  }

  // XP calculation - direct port from Java
  public static getXp(job: Job, level: number): number {
    const baseHp = this.getBaseHp(job);
    const baseMp = this.getBaseMp(job);
    const perturbation = Math.floor(2.3 * baseHp + 1.7 * baseMp);
    let value = 0;

    for (let i = 2; i <= level; i++) {
      value = value + ((13) * (i * i) / 7) + 10;
    }

    value = Math.round(value + Math.floor((perturbation * value / (190 - level))));
    return Math.floor(value);
  }

  // ATK calculation - direct port from Java
  public static getAtk(job: Job, level: number): number {
    const aMod = this.getAMod(job);
    let atk = Math.floor(aMod * 4);

    for (let i = 1; i < level; i++) {
      atk += aMod;
    }

    return Math.round(atk);
  }

  // DEF calculation - direct port from Java
  public static getDef(job: Job, level: number): number {
    const dMod = this.getDMod(job);
    let def = Math.round(dMod * 3);

    for (let i = 1; i < level; i++) {
      def += dMod;
    }

    return Math.round(def);
  }

  // Spell mapping - simplified for now (would need full PSLibSpell port)
  public static getMapLevelSpell(job: Job): Map<number, Spell[]> {
    if (!this.spellMappings.has(job)) {
      this.initLevelSpellMapping(job);
    }
    return this.spellMappings.get(job) || new Map();
  }

  private static initLevelSpellMapping(job: Job): void {
    const mapping = new Map<number, Spell[]>();

    // Direct port of Java spell mappings
    switch (job) {
      case Job.ADVENTURER:
        mapping.set(4, [SpellFactory.createSpell(PS1Spell.REST)]);
        mapping.set(5, [SpellFactory.createSpell(PS1Spell.ESCAPE)]);
        mapping.set(6, [SpellFactory.createSpell(PS1Spell.CHAT)]);
        mapping.set(8, [SpellFactory.createSpell(PS1Spell.FIRE)]); // Changed, was 12
        mapping.set(14, [SpellFactory.createSpell(PS1Spell.ROPE)]);
        mapping.set(16, [SpellFactory.createSpell(PS1Spell.FLY)]);
        mapping.set(20, [SpellFactory.createSpell(PS1Spell.CURE)]); // New
        mapping.set(26, [SpellFactory.createSpell(PS1Spell.GI_FIRE)]); // New
        mapping.set(30, [SpellFactory.createSpell(PS1Spell.ROPE_ALL)]); // New
        break;

      case Job.NATURER:
        mapping.set(6, [SpellFactory.createSpell(PS1Spell.CURE)]);
        mapping.set(9, [SpellFactory.createSpell(PS1Spell.FEAR)]);
        mapping.set(12, [SpellFactory.createSpell(PS1Spell.WALL)]);
        mapping.set(15, [SpellFactory.createSpell(PS1Spell.TRAP)]);
        mapping.set(17, [SpellFactory.createSpell(PS1Spell.EXIT)]);
        mapping.set(20, [SpellFactory.createSpell(PS1Spell.FORCE)]);
        mapping.set(22, [SpellFactory.createSpell(PS1Spell.FEAR_ALL)]); // New
        mapping.set(25, [SpellFactory.createSpell(PS1Spell.POWER_CURE)]); // New
        break;

      case Job.ESPER: // shifted by 2 some spells
        mapping.set(1, [SpellFactory.createSpell(PS1Spell.LIGHT), SpellFactory.createSpell(PS1Spell.FIRE)]);
        mapping.set(4, [SpellFactory.createSpell(PS1Spell.W_REST)]); // New
        mapping.set(7, [SpellFactory.createSpell(PS1Spell.W_CURE)]);
        mapping.set(8, [SpellFactory.createSpell(PS1Spell.EXIT)]); // 6+2
        mapping.set(10, [SpellFactory.createSpell(PS1Spell.TELE)]); // 8+2
        mapping.set(12, [SpellFactory.createSpell(PS1Spell.WIND)]);
        mapping.set(14, [SpellFactory.createSpell(PS1Spell.PROT)]);
        mapping.set(16, [SpellFactory.createSpell(PS1Spell.GI_FIRE)]); // New
        mapping.set(18, [SpellFactory.createSpell(PS1Spell.THUNDER)]);
        mapping.set(20, [SpellFactory.createSpell(PS1Spell.REVIVE)]);
        mapping.set(22, [SpellFactory.createSpell(PS1Spell.OPEN)]); // Was 17
        mapping.set(28, [SpellFactory.createSpell(PS1Spell.F_REVIVE)]); // New
        break;

      default:
        break;
    }

    this.spellMappings.set(job, mapping);
  }

  // Localized job name - direct port from Java
  public static toString(job: Job): string {
    const s = job.toString();
    const formatted = s.substring(0, 1) + s.substring(1).toLowerCase();
    return PSGame.getString(`Job_${formatted}`);
  }
}